### ISSUE TITLE
[BUG] @Listeners doesn't work for interfaces

### ISSUE THREAD

**Issue by shaburov** [cbeust/testng#2385]
### TestNG Version - 7.3.0

```java
@Listeners(IntellijIdeaTestNgPluginListener.class)
public interface IITest {
    void setTestCaseId();
}
```

### Expected behavior
The listener `IntellijIdeaTestNgPluginListener` works (enable) for all `IITest` interface implementations.

### Actual behavior
Lister does not work.

### Is the issue reproductible on runner?

- [X] Shell
- [X] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [X] IntelliJ
- [ ] NetBeans

**Comment by kobebryant2007** [comment]
Could you please show all you test code which can be reproduce your problem, thanks.

**Comment by shaburov** [comment]
@kobebryant2007 I will gladly attach again! With explanations of course...
```java
@Listeners(IntellijIdeaTestNgPluginListener.class)
public interface IITest {
    void setTestCaseId();
}
```
1. This code contains the interface `IITest` from which all test classes inherit
2. The `IITest` interface contains the `@Listeners` annotation
3. The `@Listeners` annotation contains `IntellijIdeaTestNgPluginListener.class`
4. The `IntellijIdeaTestNgPluginListener.class` implemented `IInvokedMethodListener` and contains constructor with:
```java
public final class IntellijIdeaTestNgPluginListener implements IInvokedMethodListener {
    public IntellijIdeaTestNgPluginListener() {
        System.out.println("Stop mocking and turn on your brain!!!");
        System.exit(1);
    }
}
```
When running tests of classes that inherit from the IITest interface, the listener is not created and the message from the constructor is not displayed.

**Comment by kobebryant2007** [comment]
@shaburov Thanks for your explanation. 
I tried your test case and I found that the listeners only works when you add them in implementation classes. I think the @Listeners annotation is not @Inherited, so the annotation can not be inherited in test implementation class.
@krmahadevan should we support this feature?

**Comment by shaburov** [comment]
@krmahadevan @kobebryant2007 This is a request for fix.
If you use annotations, then you are obliged to find the annotations for both classes and interfaces. And I doubt this is just a feature. This is a clear implementation bug.

![image](https://user-images.githubusercontent.com/15269131/96362250-ab9b8600-1134-11eb-92b4-1e33cab48a84.png)

**Comment by krmahadevan** [comment]
@shaburov 

> If you use annotations, then you are obliged to find the annotations for both classes and interfaces. And I doubt this is just a feature. This is a clear implementation bug.

I don't remember seeing any such statement anywhere within the TestNG documentation that states that `@Listeners` should/would work when annotated on listeners. There are some gaps in how this should work.
A user may simply add the annotation to an interface which never gets implemented by any class. in that situation there's nothing that TestNG can do in terms of figuring out if this listener should be wired in. 

So IMO its not a bug, but can be treated as a feature request. 

Since the following is not clear, I would want to get inputs from both @cbeust and @juherr 

* Is this a realistic use-case that a regular user would do ? Listeners are at the end of the day "global" in nature and are never scoped at "class" level. So why should a user have to resort to annotating an interface with this annotation ? The listener can always be built such that it recognizes an object type (if your idea of having an interface is to define a common type for a test class) and then inspect the test class instance type to find out if the listener is to be applied or not.
* If this is a realistic use-case then I would like to understand what would/should be the expected behavior, when the annotation is added on top of a listener which never gets applied. 

In a nutshell, I would need to have a bit more clarity on the expectations here.

In the meantime, if you have an implementation or an approach in mind, please feel free to submit a PR and we can help you get that vetted out as well.

@cbeust @juherr - WDYT ?

**Comment by juherr** [comment]
My feeling is that most of the frameworks have their annotation `@Inherited` and it is more natural for me too.

I don't see any reason to not implement it and I think it will just `add` some other listeners.

**Comment by kobebryant2007** [comment]
OK, I will do some investigation and maybe raise a PR later.
One thing I should reminder again is that @Inherited only works when annotate a class, and the interface doesn't support inherited. 
so we should use other way.

**Comment by kobebryant2007** [comment]
Hi @krmahadevan 
Based on your comments, could you please specify the expect behavior of 5 and 6? Listenners in different class or interface should be merged or override?
Please also notice the comments in findAnnotationInSuperClasses , Listenners annotation would be handled specially for ignore aggregation.

**Comment by krmahadevan** [comment]
@kobebryant2007 

> Based on your comments, could you please specify the expect behavior of 5 and 6? Listenners in different class or interface should be merged or override?

```
5. Can you please add a test that checks the behaviour when there is inheritance of interfaces involved (multi level and multiple inheritance both) and every interface attempts at wiring in the same listener.
```

TestNG already gives a guarantee that duplicate instances of the same listener will not be wired in (They may get instantiated but they wont get hooked in). So this needs to be met for this case.

```
6. Can you please add a test that checks the behaviour when there is inheritance of interfaces involved (multi level and multiple inheritance both) and every interface attempts at wiring unique listeners.
```
Here the expectation is that all listeners get wired in.

> Please also notice the comments in findAnnotationInSuperClasses , Listenners annotation would be handled specially for ignore aggregation.

The comment [here](https://github.com/cbeust/testng/blob/ae6d2fdf5628e0457dc507cd5619954def73e727/src/main/java/org/testng/internal/annotations/JDK15AnnotationFinder.java#L77-L79) shouldn't have an impact here, because all we are doing is, we are aggregating listeners while walking up the inheritance hierarchy. We just need to ensure that there are no duplicate instances of the same listener. As long as that expectation is met, I guess we should be fine.

**Comment by krmahadevan** [comment]
This issue stands closed via https://github.com/cbeust/testng/pull/2398
